witness/ethereum/contracts/RootWitness.sol

```diff
@@ -0,0 +1,229 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.0;
+
+import {WitnessUtils} from "./WitnessUtils.sol";
+
+/// @author sinasabet@google.com
+/// @title Witness for tracking external merkle trees using consistency proofs.
+contract RootWitness is WitnessUtils {
+    /// @dev struct to encapsulate data for a specific tree.
+    struct TreeData {
+        bytes32 root;
+        uint256 size;
+        uint256 timestamp;
+        bytes32 signature;
+    }
+
+    /// @dev struct to encapsulate params of a single update call
+    struct UpdateData {
+      bytes32 treeId;
+      bytes32[] proof;
+      uint256 size;
+      uint256 timestamp;
+      bytes32 signature;
+    }
+
+    /// @dev emits information related to an update method call.
+    event Updated(
+        bytes32 treeId,
+        TreeData newTreeData,
+        bytes32[] consistencyProof
+    );
+
+    address public owner;
+    /// @dev mapping from treeId to that tree's data.
+    mapping(bytes32 => TreeData) public treeDatas;
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    /// @dev getter for entire treeData value of a specific treeId.
+    function getTreeDataById(bytes32 treeId)
+        public
+        view
+        returns (TreeData memory)
+    {
+        return treeDatas[treeId];
+    }
+
+    /// @dev handler for batch updates.
+    function batchUpdate(UpdateData[] calldata updateDatas)
+      external
+      returns
+      (TreeData[] memory)
+    {
+      TreeData[] memory results = new TreeData[](updateDatas.length);
+      for (uint256 i = 0; i < updateDatas.length; i++) {
+        UpdateData calldata data = updateDatas[i];
+        results[i] = update(
+          data.treeId,
+          data.proof,
+          data.size,
+          data.timestamp,
+          data.signature
+        );
+      }
+      return results;
+    }
+
+    /// Handles an update for the witness for a specific treeId.
+    /// @param treeId the id of the tree the update is for.
+    /// @param consistencyProof list of hashes making up the consistency proof.
+    /// @param size the size of the updated tree.
+    ///   Corresponds to end index of the right range.
+    /// @param timestamp integer timestamp of the update.
+    /// @param signature raw signature over the marshalled bytes of new root.
+    /// @dev merge rightRange into compactRange.
+    /// @return treeData updated treeData entry for treeId.
+    function update(
+        bytes32 treeId,
+        bytes32[] calldata consistencyProof,
+        uint256 size,
+        uint256 timestamp,
```

Comment:
Are the timestamps possibly optimizable in terms of size?

A `uint32` gives enough room for Unix epoch timestamps in case this is the format desired. Otherwise `uint40` is more than enough.